File: goberkeley/src/berkeley/socket_zero_mq_exception.go
package berkeley

import (
	"fmt"
)

// ErrorType enum para los diferentes tipos de errores posibles en ZeroMQ.
type ErrorType int

const (
	// CONNECTION_ERROR Error relacionado con la conexión de ZeroMQ
	CONNECTION_ERROR ErrorType = iota
	// SEND_ERROR Error al enviar un mensaje a través del socket
	SEND_ERROR
	// RECEIVE_ERROR Error al recibir un mensaje del socket
	RECEIVE_ERROR
)

// SocketZeroMQException estructura que encapsula un error personalizado para ZeroMQ.
type SocketZeroMQException struct {
	Message   string    // Mensaje de error
	ErrorType ErrorType // Tipo de error
}

// NewSocketZeroMQException crea una nueva instancia de SocketZeroMQException.
func NewSocketZeroMQException(message string, errorType ErrorType) *SocketZeroMQException {
	return &SocketZeroMQException{
		Message:   message,
		ErrorType: errorType,
	}
}

// Error implementación de la interfaz error para SocketZeroMQException.
func (e *SocketZeroMQException) Error() string {
	return fmt.Sprintf("Error: %s, Tipo de error: %v", e.Message, e.ErrorType)
}

// GetErrorType obtiene el tipo de error.
func (e *SocketZeroMQException) GetErrorType() ErrorType {
	return e.ErrorType
}

File: jberkeley/src/org/jogugi/berkeley/SocketZeroMQException.java
package org.jogugi.berkeley;

/**
 * @brief Excepción personalizada para errores en la operación de sockets de ZeroMQ.
 *
 * Esta clase extiende de `Exception` y se utiliza para manejar errores específicos que ocurren
 * durante las operaciones de sockets utilizando ZeroMQ. Los tipos de errores cubiertos son
 * relacionados con la conexión, el envío y la recepción de mensajes.
 */
public class SocketZeroMQException extends Exception {

    /**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	/**
     * @brief Enumeración de los tipos de errores posibles.
     *
     * Esta enumeración se utiliza para clasificar los diferentes tipos de errores que pueden ocurrir
     * durante las operaciones con sockets ZeroMQ.
     */
    public enum ErrorType {
        CONNECTION_ERROR, /**< Error relacionado con la conexión de ZeroMQ. */
        SEND_ERROR,       /**< Error al enviar un mensaje a través del socket. */
        RECEIVE_ERROR     /**< Error al recibir un mensaje del socket. */
    }

    private ErrorType errorType; /**< Tipo de error que ocurrió. */

    /**
     * @brief Constructor de la excepción.
     *
     * @param message El mensaje de error que describe el problema ocurrido.
     * @param errorType El tipo de error que ocurrió, basado en la enumeración `ErrorType`.
     */
    public SocketZeroMQException(String message, ErrorType errorType) {
        super(message); // Llamada al constructor de la clase base (Exception)
        this.errorType = errorType;
    }

    /**
     * @brief Obtiene el tipo de error asociado a esta excepción.
     *
     * @return El tipo de error que se produjo durante la operación de ZeroMQ.
     */
    public ErrorType getErrorType() {
        return errorType;
    }
}